File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/MoraleIndicator.vb
Namespace Controls
    Public NotInheritable Class MoraleIndicator
        Inherits Control

        Public Sub New()
            Me.DefaultStyleKey = GetType(MoraleIndicator)
        End Sub

        Public ReadOnly MoraleProperty As DependencyProperty =
            DependencyProperty.Register(NameOf(Morale), GetType(Integer), GetType(MoraleIndicator),
                                        New PropertyMetadata(0, Sub(d, e) DirectCast(d, MoraleIndicator).CheckVisualState()))
        Public Property Morale As Integer
            Get
                Return DirectCast(GetValue(MoraleProperty), Integer)
            End Get
            Set(value As Integer)
                SetValue(MoraleProperty, value)
            End Set
        End Property

        Protected Overrides Sub OnApplyTemplate()
            MyBase.OnApplyTemplate()
            CheckVisualState()
        End Sub

        Private Sub CheckVisualState()
            Select Case Morale
                Case > 85
                    VisualStateManager.GoToState(Me, "Full", True)
                Case > 49
                    VisualStateManager.GoToState(Me, "High", True)
                Case > 39
                    VisualStateManager.GoToState(Me, "Normal", True)
                Case > 29
                    VisualStateManager.GoToState(Me, "Slight", True)
                Case > 19
                    VisualStateManager.GoToState(Me, "Moderate", True)
                Case Else
                    VisualStateManager.GoToState(Me, "Serious", True)
            End Select
        End Sub
    End Class
End Namespace

File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/MoraleIndicator.cs
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;

namespace Sakuno.ING.Views.UWP.Controls
{
    public sealed class MoraleIndicator : Control
    {
        public MoraleIndicator()
        {
            DefaultStyleKey = typeof(MoraleIndicator);
        }

        public static readonly DependencyProperty MoraleProperty
            = DependencyProperty.Register(nameof(Morale), typeof(int), typeof(MoraleIndicator),
                new PropertyMetadata(0, (d, e) => ((MoraleIndicator)d).CheckVisualState()));
        public int Morale
        {
            get => (int)GetValue(MoraleProperty);
            set => SetValue(MoraleProperty, value);
        }

        protected override void OnApplyTemplate()
        {
            base.OnApplyTemplate();
            CheckVisualState();
        }

        private void CheckVisualState()
        {
            int morale = Morale;
            if (morale > 85)
                VisualStateManager.GoToState(this, "Full", true);
            else if (morale > 49)
                VisualStateManager.GoToState(this, "High", true);
            else if (morale > 39)
                VisualStateManager.GoToState(this, "Normal", true);
            else if (morale > 29)
                VisualStateManager.GoToState(this, "Slight", true);
            else if (morale > 19)
                VisualStateManager.GoToState(this, "Moderate", true);
            else
                VisualStateManager.GoToState(this, "Serious", true);
        }
    }
}
